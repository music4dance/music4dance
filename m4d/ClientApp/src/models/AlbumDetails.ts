import { jsonMember, jsonObject } from "typedjson";
import { PurchaseEncoded } from "./Purchase";

@jsonObject
export class AlbumDetails {
  @jsonMember(Number) public index?: number;
  @jsonMember(String) public name?: string;
  @jsonMember(Number) public track?: number;
  @jsonMember(String) public publisher?: string;
  @jsonMember(PurchaseEncoded) public purchase!: PurchaseEncoded;

  public constructor(init?: Partial<AlbumDetails>) {
    Object.assign(this, init);
    if (!this.purchase) {
      this.purchase = new PurchaseEncoded();
    }
  }
}
