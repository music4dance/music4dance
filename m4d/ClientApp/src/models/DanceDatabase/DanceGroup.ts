import { jsonArrayMember, jsonMember, jsonObject } from "typedjson";
import { NamedObject } from "./NamedObject";
import type { DanceType } from "./DanceType";
import { assign } from "@/helpers/ObjectHelpers";

@jsonObject
export class DanceGroup extends NamedObject {
  @jsonMember(String) blogTag?: string;
  @jsonArrayMember(String) public danceIds!: string[];
  public dances: DanceType[] = [];

  // TODO: Lookes like TypedJSON doesn't hydrate the object
  //  in a way that instanceof works, so we'll use a different method
  //   public static isGroup(obj: NamedObject): obj is DanceGroup {
  //     return obj instanceof DanceGroup;
  //   }

  public static isGroup(obj: NamedObject): obj is DanceGroup {
    if (!obj) {
      return false;
    }
    return (obj as DanceGroup).danceIds !== undefined;
  }

  public constructor(init?: Partial<DanceGroup>) {
    super();
    assign(this, init);
  }
}
