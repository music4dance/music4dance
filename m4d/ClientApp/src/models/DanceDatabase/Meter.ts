import { jsonMember, jsonObject } from "typedjson";

@jsonObject
export class Meter {
  @jsonMember(Number) public numerator!: number;
  @jsonMember(Number) public denominator!: number;

  constructor(numerator = 0, denominator = 0) {
    this.numerator = numerator;
    this.denominator = denominator;
  }

  public equals(that: Meter): boolean {
    return this.numerator === that.numerator && this.denominator === that.denominator;
  }

  public toString(): string {
    return `${this.numerator}/${this.denominator}`;
  }

  public static get EmptyMeter(): Meter {
    const meter = new Meter();
    meter.numerator = 0;
    meter.denominator = 0;
    return meter;
  }
}
