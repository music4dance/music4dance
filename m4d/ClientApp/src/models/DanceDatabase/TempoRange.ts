import { jsonMember, jsonObject } from "typedjson";

@jsonObject
export class TempoRange {
  @jsonMember(Number) public min!: number;
  @jsonMember(Number) public max!: number;

  constructor(min = 0, max = 0) {
    this.min = min;
    this.max = max;
  }

  public toString(separator = "-"): string {
    return (
      this.formatTempo(this.min) +
      (this.min === this.max ? "" : separator + this.formatTempo(this.max))
    );
  }

  public bpm(numerator: number = 1, separator = "-"): string {
    return (
      this.formatTempo(this.min * numerator) +
      (this.min === this.max ? "" : separator + this.formatTempo(this.max * numerator))
    );
  }

  public formatBPM(numerator = 1, separator = "-"): string {
    return `${this.bpm(numerator, separator)} BPM`;
  }

  public mpm(numerator: number = 1, separator = "-"): string {
    return (
      this.formatTempo(this.min / numerator) +
      (this.min === this.max ? "" : separator + this.formatTempo(this.max / numerator))
    );
  }

  public formatMPM(numerator = 1, separator = "-"): string {
    return `${this.bpm(numerator, separator)} MPM`;
  }

  public calculateDelta(tempo: number): number {
    return tempo > this.max ? tempo - this.max : tempo < this.min ? tempo - this.min : 0;
  }

  public calculateDeltaPercent(tempo: number) {
    return (this.calculateDelta(tempo) * 100) / (tempo >= this.max ? this.max : this.min);
  }

  public include(that?: TempoRange | null): TempoRange {
    return that ? new TempoRange(Math.min(this.min, that.min), Math.max(this.max, that.max)) : this;
  }

  public get isInfinite(): boolean {
    return this.min < 10 && this.max > 499;
  }

  public toBpm(numerator: number): TempoRange {
    return new TempoRange(this.min * numerator, this.max * numerator);
  }

  public toMpm(numerator: number): TempoRange {
    return new TempoRange(this.min / numerator, this.max / numerator);
  }

  private formatTempo(tempo: number): string {
    return parseFloat(tempo.toFixed(1)).toString();
  }
}
