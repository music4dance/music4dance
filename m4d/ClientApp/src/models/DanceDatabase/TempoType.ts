export enum TempoType {
  None = 0,
  Beats = 1,
  Measures = 2,
  Both = 3,
}
