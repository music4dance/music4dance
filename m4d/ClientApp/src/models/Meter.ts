import { jsonMember, jsonObject } from "typedjson";

@jsonObject
export class Meter {
  @jsonMember(Number) public numerator!: number;
  @jsonMember(Number) public denominator!: number;

  constructor(numerator = 0, denominator = 0) {
    this.numerator = numerator;
    this.denominator = denominator;
  }

  public toString(): string {
    return `${this.numerator}/${this.denominator}`;
  }

  public static get EmptyMeter(): Meter {
    const meter = new Meter();
    meter.numerator = 0;
    meter.denominator = 0;
    return meter;
  }
}
