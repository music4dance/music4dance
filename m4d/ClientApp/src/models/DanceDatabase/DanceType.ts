import { kebabToWords, wordsToKebab } from "@/helpers/StringHelpers";
import { jsonArrayMember, jsonMember, jsonObject } from "typedjson";
import { DanceInstance } from "./DanceInstance";
import { DanceObject } from "./DanceObject";
import { TempoRange } from "./TempoRange";
import { assign } from "@/helpers/ObjectHelpers";
import type { DanceGroup } from "./DanceGroup";

@jsonObject({ onDeserialized: "onDeserialized" })
export class DanceType extends DanceObject {
  @jsonArrayMember(String) public organizations: string[] = [];
  @jsonMember(String) public link!: string;
  @jsonArrayMember(DanceInstance) public instances: DanceInstance[] = [];
  public groups: DanceGroup[] = [];

  public constructor(init?: Partial<DanceType>) {
    super();
    if (init) {
      assign(this, init);
      this.onDeserialized();
    }
  }

  private onDeserialized(): void {
    if (this.instances) {
      for (const inst of this.instances) {
        inst.danceType = this;
      }
    }
  }

  public reduce(instances: DanceInstance[]): DanceType {
    const other = new DanceType(this);
    other.instances = [...instances];
    for (const instance of other.instances) {
      instance.danceType = other;
    }

    return other;
  }

  public get styles(): string[] {
    return this.instances.map((inst) => inst.style);
  }

  public get competitionDances(): DanceInstance[] {
    return this.instances.filter((inst) => inst.competitionGroup);
  }

  public get tempoRange(): TempoRange {
    return this.instances.map((d) => d.tempoRange).reduce((acc, inst) => acc.include(inst));
  }

  public inGroup(group: string): boolean {
    return this.groups.find((g) => g.name === group) !== undefined;
  }

  public filteredStyles(filter: string[]): string[] {
    return filter ? this.styles.filter((s) => filter.indexOf(wordsToKebab(s)) !== -1) : this.styles;
  }

  public filteredTempo(styles: string[], organizations: string[]): TempoRange | undefined {
    if (!styles.length && !organizations.length) {
      return this.tempoRange;
    }

    let ret: TempoRange | undefined;
    for (const inst of this.instances) {
      if (
        !inst.tempoRange ||
        (styles.length && !styles.find((s) => s === wordsToKebab(inst.style)))
      ) {
        continue;
      }

      const newRange = inst.filteredTempo(organizations);
      if (!ret) {
        ret = newRange;
      } else if (newRange) {
        ret = ret.include(newRange);
      }
    }

    return ret;
  }

  private instanceFromStyle(style: string): DanceInstance | undefined {
    return this.instances.find((i) => i.style === kebabToWords(style));
  }

  // TODO: should we throw out the whole thing if the dancetype isn't specified by the organization?
  private buildOrganizationIds(organizations: string[]): string[] {
    const prefixes = ["Social", "DanceSport", "NDCA"];
    return prefixes.filter((p) => organizations.find((o) => o.indexOf(p) !== -1));
  }
}
