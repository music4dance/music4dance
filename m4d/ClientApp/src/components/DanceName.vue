<script setup lang="ts">
import { TempoType } from "@/models/DanceDatabase/TempoType";
import { DanceType } from "@/models/DanceDatabase/DanceType";
import type { NamedObject } from "@/models/DanceDatabase/NamedObject";
import { computed } from "vue";
import { DanceGroup } from "@/models/DanceDatabase/DanceGroup";

const props = withDefaults(
  defineProps<{
    dance: NamedObject;
    showTempo?: TempoType;
    showSynonyms?: boolean;
    multiLine?: boolean;
    hideLink?: boolean;
  }>(),
  {
    showTempo: TempoType.None,
    showSynonyms: false,
    multiLine: false,
    hideLink: false,
  },
);

const danceLink = computed(() => {
  return `/dances/${props.dance.seoName}`;
});

const canShowTempo = computed(() => {
  const dance = DanceGroup.isGroup(props.dance) ? undefined : (props.dance as DanceType);
  return !!dance && !dance.tempoRange.isInfinite;
});

const tempoText = computed(() => {
  const dance = DanceGroup.isGroup(props.dance) ? undefined : (props.dance as DanceType);
  if (!dance) return "";

  const showTempo = props.showTempo;
  const bpm = showTempo & TempoType.Beats ? `${dance.tempoRange.toString()} BPM` : "";
  const mpm =
    showTempo & TempoType.Measures ? `${dance.tempoRange.mpm(dance.meter.numerator)} MPM` : "";

  return `${bpm}${mpm && bpm ? "/" : ""}${mpm}`;
});

const synonymText = computed(() => {
  const synonyms = props.dance.synonyms;
  return props.showSynonyms && synonyms ? `${synonyms.join(", ")}` : "";
});

// Exposed for testing
defineExpose({ danceLink, canShowTempo, tempoText, synonymText });
</script>

<template>
  <span
    ><span v-if="hideLink">{{ dance.name }}</span
    ><a v-else :href="danceLink">{{ dance.name }}</a>
    <span v-if="synonymText"
      ><br v-if="multiLine" />
      ({{ synonymText }})
    </span>
    <span v-if="showTempo && canShowTempo" style="font-size: 0.8rem" class="ms-2">
      {{ tempoText }}</span
    >
  </span>
</template>
