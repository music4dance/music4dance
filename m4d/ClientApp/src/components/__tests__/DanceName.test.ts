import { describe, expect, test } from "vitest";
import { mount } from "@vue/test-utils";
import DanceName from "@/components/DanceName.vue";
import { TempoType } from "@/models/TempoType";
import { DanceGroup } from "@/models/DanceGroup";
import { DanceType } from "@/models/DanceType";
import { TempoRange } from "@/models/TempoRange";
import { Meter } from "@/models/Meter";

const dance = new DanceType({
  name: "test-dance",
  tempoRange: new TempoRange(100, 110),
  meter: new Meter(4, 4),
});
const infiniteDance = new DanceType({ name: "test-dance", tempoRange: new TempoRange(0, 10000) });

describe("DanceName.vue", () => {
  test("computes danceLink correctly", () => {
    const wrapper = mount(DanceName, {
      props: {
        dance: dance,
      },
    });

    // @ts-ignore:next-line
    expect(wrapper.vm.danceLink).toBe("/dances/test-dance");
  });

  test("computes canShowTempo correctly for finite tempo", () => {
    const wrapper = mount(DanceName, {
      props: {
        dance: dance,
      },
    });

    // @ts-ignore:next-line
    expect(wrapper.vm.canShowTempo).toBe(true);
  });

  test("computes canShowTempo correctly for infinite tempo", () => {
    const wrapper = mount(DanceName, {
      props: {
        dance: infiniteDance,
      },
    });

    // @ts-ignore:next-line
    expect(wrapper.vm.canShowTempo).toBe(false);
  });

  test("computes canShowTempo correctly for dance group", () => {
    const wrapper = mount(DanceName, {
      props: {
        dance: new DanceGroup({ name: "test-group", danceIds: ["test"] }),
      },
    });

    // @ts-ignore:next-line
    expect(wrapper.vm.canShowTempo).toBe(false);
  });

  test("computes tempoText correctly", () => {
    const wrapper = mount(DanceName, {
      props: {
        dance: dance,
        showTempo: TempoType.Both,
      },
    });

    // @ts-ignore:next-line
    expect(wrapper.vm.tempoText).toBe("100-110 BPM/25-27.5 MPM");
  });

  test("renders a dance name correctly without tempo", () => {
    const wrapper = mount(DanceName, {
      props: {
        dance: dance,
        showTempo: TempoType.None,
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  test("renders a dance name correctly with tempo", () => {
    const wrapper = mount(DanceName, {
      props: {
        dance: dance,
        showTempo: TempoType.Both,
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  test("renders a dance name correctly with multiline", () => {
    const wrapper = mount(DanceName, {
      props: {
        dance: dance,
        showTempo: TempoType.Both,
        multiLine: true,
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
