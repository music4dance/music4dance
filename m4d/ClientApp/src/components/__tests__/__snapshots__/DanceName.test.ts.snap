// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DanceName.vue > renders a dance name correctly with multiline 1`] =
  `"<span><a href="/dances/test-dance">test-dance</a><!--v-if--><span style="font-size: 0.8rem;" class="ms-2">100-110 BPM/25-27.5 MPM</span></span>"`;

exports[`DanceName.vue > renders a dance name correctly with tempo 1`] =
  `"<span><a href="/dances/test-dance">test-dance</a><!--v-if--><span style="font-size: 0.8rem;" class="ms-2">100-110 BPM/25-27.5 MPM</span></span>"`;

exports[`DanceName.vue > renders a dance name correctly without tempo 1`] =
  `"<span><a href="/dances/test-dance">test-dance</a><!--v-if--><!--v-if--></span>"`;
